gradle-guide-error-prone/src/main/java/com/palantir/gradle/guide/errorprone/ConfigurationAvoidanceRegistration.java

```diff
@@ -0,0 +1,136 @@
+import com.sun.source.tree.BlockTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.IdentifierTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.ReturnTree;
+import com.sun.source.tree.VariableTree;
+import com.sun.source.util.TreePath;
+import com.sun.source.util.TreePathScanner;
+import java.util.Map;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        severity = SeverityLevel.ERROR,
+        summary = "When registering a new `Task`, `Configuration` or other Gradle domain type, "
+                + "use `.register` instead of `.create` to avoid realising the object eagerly "
+                + "and performing unnecessary work which will slow down the build.")
+public final class ConfigurationAvoidanceRegistration extends GradleGuideBugChecker
+        implements BugChecker.MethodInvocationTreeMatcher {
+    private static final Matcher<ExpressionTree> MATCHER = Matchers.instanceMethod()
+            .onDescendantOfAny("org.gradle.api.NamedDomainObjectContainer")
+            .namedAnyOf("create");
+
+    private static final Matcher<MethodInvocationTree> FIRST_ARGUMENT_IS_MAP =
+            Matchers.argument(0, Matchers.isSubtypeOf(Map.class));
+
+    private static final Matcher<MethodInvocationTree> SECOND_ARGUMENT_IS_GROOVY_CLOSURE =
+            Matchers.argument(1, Matchers.isSubtypeOf("groovy.lang.Closure"));
+
+    private static final Matcher<MethodInvocationTree> NO_DIRECT_REGISTER_EQUIVALENT =
+            Matchers.anyOf(FIRST_ARGUMENT_IS_MAP, SECOND_ARGUMENT_IS_GROOVY_CLOSURE);
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+        if (!MATCHER.matches(tree, state)) {
+            return Description.NO_MATCH;
+        }
+
+        Description.Builder description = buildDescription(tree);
+        SuggestedFixBuilder fix = new SuggestedFixBuilder();
+
+        if (!bestEffortModeEnabled(state)) {
+            return description.build();
+        }
+
+        // If the first argument is a map, or second is a Closure, there isn't an equivalent
+        // `.register` method to move to (from Java code at least)
+        if (NO_DIRECT_REGISTER_EQUIVALENT.matches(tree, state)) {
+            return description.build();
+        }
+
+        TreePath parentPath = state.getPath().getParentPath();
+        if (parentPath.getLeaf() instanceof VariableTree variableTree) {
+            replaceVariableDeclartionTypeWithProvider(state, fix, variableTree);
+            replaceVariableUsagesWithTaskProviderGet(fix, parentPath);
+        }
+
+        fix.replace(
+                tree.getMethodSelect(),
+                state.getSourceForNode(ASTHelpers.getReceiver(tree.getMethodSelect())) + ".register");
+
+        if (parentPath.getLeaf() instanceof ExpressionTree || parentPath.getLeaf() instanceof ReturnTree) {
+            fix.postfixWith(tree, ".get()");
+        }
```

Comment:
any way you can replace this with a switch? a bit easier to see the control flow

Reply:
I don't think there's a nice way to do this without jdk 21 support.

Reply:
Without adding in casting at least

Reply:
Put the instanceofs next to each other at least